sips/pending/_posts/2016-09-09-inline-meta.md

```diff
@@ -0,0 +1,1023 @@
+users.map(u => u.name)
+
+// error: too many arguments for
+// method apply: (index: Int)Char in class StringOps
+//    users.map(u => u.name)
+//             ^
+```
+
+If the macro user accidentally defines a term called `Select` or `Ref`, our macro is going to stop working
+with what looks like a nonsensical error message. Because principled tool support wasn't among the design goals of
+our macro system, a macro user getting this error is mostly helpless apart from trying to use internal compiler options
+that print all macro expansions and then going through the resulting wall of text.
+
+One reliable approach to prevent hygiene errors is to use fully-qualified names for external references and
+to generate unique names for local variables. A somewhat more concise, but potentially much more laborious approach is to explicitly
+assign symbols to references and definitions emitted in macro expansions.
+This approach often requires extensive knowledge of compiler internals, so it is less frequent in the wild.
+Common to both of these approaches is that they require explicit attention from metaprogrammers and
+failures to apply them typically go unnoticed.
+
+To sum it up, even in this simple macro we encountered situations where knowledge of compiler internals
+(the desugaring of the `QueryApi` application, the fact that non-alphanumeric names are encoded) was essential.
+We also ran into problems with tool support and hygiene, which demonstrates how important they are to a macro system.
+These are all common criticisms of def macros.
+
+### Inline/meta
+
+User interface of def macros is a seamless extension to the existing language interface.
+Thanks to macro defs looking like regular methods and macro applications looking like regular method applications,
+macro users are likely to not even realize that they are using macros.
+
+Unfortunately, metaprogrammer interface leaves much to be desired. First, in the current macro system,
+metaprograms have to be defined separately from their signatures, typically involving helper objects and
+duplication of parameters. Secondly, the underlying metaprogramming API requires
+knowing bits and pieces of compiler internals. For example, in the case of `Query.map`,
+the metaprogrammer had to know the internal representation of the `QueryApi` application
+and internal details of how names are represented.
+
+New-style macros provide the same user interface, and at the same time significantly improve metaprogrammer
+interface by enabling lightweight macro declaration syntax and using a better metaprogramming API.
+
+```
+object Query {
+  implicit class QueryApi[T](q: Query[T]) {
+    inline def map[U](fn: T => U): Query[U] = meta {
+      import scala.meta._
+
+      val q"$_($prefix)" = this
+
+      val node: Tree = fn match {
+        case q"($name: $_) => $body" =>
+          body match {
+            case q"$qual.$_" if qual =:= name =>
+              q"Ref[${body.tpe}](${name.toString})"
```

Comment:
Why is the example using `${name.toString}` in the constructor argument of Ref? As far as I understand, this should be name of the parameter of the lambda, i.e. `u`. Consequently, the resulting code would be `Ref[String]("u")`. 

For the sake of the example I think the desired expansion would be `Ref[String]("name")`, so the macro would have to capture the name of the selected field in the inner-most case-clause.